streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -154,7 +155,13 @@ private void registerMetrics() {
         StateStoreMetrics.addNumOpenIteratorsGauge(taskId.toString(), metricsScope, name(), streamsMetrics,
                 (config, now) -> numOpenIterators.sum());
         StateStoreMetrics.addOldestOpenIteratorGauge(taskId.toString(), metricsScope, name(), streamsMetrics,
-                (config, now) -> openIterators.isEmpty() ? null : openIterators.first().startTimestamp()
+                (config, now) -> {
+                    try {
+                        return openIterators.isEmpty() ? null : openIterators.first().startTimestamp();
```

Comment:
Not sure if we need to try-catch... What about using `pollFirst()` ?

Reply:
Unfortunately, that method removes the element from the Set, which we don't want to do in the Gauge!

Reply:
Ah. Good point. Missed this...

For this case, does it even make sense to check for `isEmpty()`? We would also just to `return openIterators.first().startTimestamp();` only?

Reply:
Or we need to use `synchronized(openIterators)` everywhere... -- I don't think we are on the hot code path, so might be ok?

Reply:
\cc @ableegoldman @cadonna @lucasbru @bbejeck @aliehsaeedii for feedback? No solution seems to be satisfying.

Reply:
@mjsax I know it's a bit ugly, but I think this is the best we can do.

`synchronized (openIterators)` is definitely worse, especially in the presence of multiple query threads, but even with just a stream thread, the metrics reporting thread would impact its performance. Throwing an Exception may not be massively fast, but it's only impacting the metrics reporting thread, and we only throw one in the rare case that there's a race between the `isEmpty()` check and `first()`.

> For this case, does it even make sense to check for isEmpty()? We would also just to return openIterators.first().startTimestamp(); only?

We still want to check `isEmpty()` first, simply to avoid having to throw the Exception in the common case that the Set is empty. The Exception becomes a guard against a race between the `isEmpty()` and `first()` invocations, which are pretty rare, but occasionally happen.

> Or use pollFirst() and re-insert afterwards (also not good code TBH...)

This would cause a race with the stream threads/query threads, because when the Iterator is closed, they remove it from the `openIterators` Set. If this happens while a metrics reporter thread has temporarily removed the Iterator from the Set, it would fail to be removed, and then the reporter thread would add it back, causing the Iterator to leak!

Reply:
Actually, another option is to use `ConcurrentSkipListMap<MeteredIterator, Object>`. This is actually what `ConcurrentSkipListSet` uses internally, and `ConcurrentSkipListMap` provides `firstEntry()`, which returns `null` when empty.

The downside of this is that everwhere that adds to `openIterators` would need to use a dummy value (`ConcurrentSkipListSet` uses `Boolaen.TRUE`), which might look a bit odd.

Reply:
Maybe you can just write your own version of `ConcurrentSkipListSet`  for these iterators, that wraps the `ConcurrentSkipListMap` while the providing the needed APIs, and hides all the dummy value stuff so it doesn't get too messy in the code itself?

I honestly don't think the current proposal (just catching the exceptions) is that bad, but I understand wanting something better

Reply:
> Maybe you can just write your own version of ConcurrentSkipListSet for these iterators, that wraps the ConcurrentSkipListMap while the providing the needed APIs, and hides all the dummy value stuff so it doesn't get too messy in the code itself?

Reply:
Thanks for the details/background Nick.

If nobody objects, I can go ahead a merge this PR as-is in the next days (just waiting to give people so more time to chime in).

Reply:
> Unfortunately, the internals of ConcurrentSkipListSet are private, so our own version would need to be a complete copy of the JDK implementation. This is probably not a good idea, as upgrading JDK would not benefit from any improvements in the JDK implementation.

I actually think this wouldn't be so bad, since as you pointed out, the ConcurrentSkipListSet is basically a no-op wrapper around the ConcurrentSkipListMap which is where all the messy stuff happens. You could more or less copy the existing ConcurrentSkipListSet implementation as-is and just change the name (and add the `#firstEntry` API we need), it truly does nothing but delegate to the underlying wrap. So I can't imagine there would be anything to change, much less improve, in the JDK implementation of  ConcurrentSkipListSet 

Just wanted to clarify, as I said before I'm actually fine with the way it's implemented now so I'm not asking for anything to be changed. But if we ever do want to change this or run into a similar problem, imo writing our own wrapper around the ConcurrentSkipListMap is the way to go 🤷‍♀️